File: sdks/typescript/src/ai/provider-openai.ts
/**
 * Dunno-wrapped OpenAI client.
 * Usage: import { OpenAI } from '@dunno/dunno/ai/provider-openai'
 */
import type { DunnoClientOptions } from '../types.js';
import { DunnoClient } from '../client.js';

let _defaultClient: DunnoClient | null = null;

function getDefaultClient(options?: DunnoClientOptions): DunnoClient {
  if (!_defaultClient) _defaultClient = new DunnoClient(options);
  return _defaultClient;
}

export function createDunnoOpenAI(options?: DunnoClientOptions) {
  return (BaseOpenAI: any) => {
    const dunno = getDefaultClient(options);

    return class DunnoOpenAI extends BaseOpenAI {
      constructor(...args: any[]) {
        super(...args);
        const baseChatCompletions = (this as any).chat.completions;

        const wrappedCreate = async (params: any) => {
          const {
            dunnoAgent,
            dunnoSession,
            dunnoPerson,
            dunnoAgentVersion,
            ...rest
          } = params;

          const start = Date.now();
          const result = await baseChatCompletions.create(rest);
          const latencyMs = Date.now() - start;

          if (dunnoSession) {
            dunno.events.create({
              eventName: 'llm',
              properties: {
                model: result.model,
                inputTokens: result.usage?.prompt_tokens,
                outputTokens: result.usage?.completion_tokens,
                latencyMs,
                messages: rest.messages,
              },
              dunnoSession,
              dunnoAgent,
              dunnoAgentVersion,
              dunnoPerson,
            });
          }

          return result;
        };

        (this as any).chat = {
          completions: { create: wrappedCreate },
        };
      }
    };
  };
}

// Convenience: auto-wrap if openai is installed
export async function OpenAI(options?: DunnoClientOptions & { openaiOptions?: Record<string, unknown> }) {
  const { default: BaseOpenAI } = await import('openai');
  const Wrapped = createDunnoOpenAI(options)(BaseOpenAI);
  return new Wrapped(options?.openaiOptions);
}

// Static class form for synchronous usage
export class OpenAIClass {
  private _inner: any;
  private _dunno: DunnoClient;

  constructor(opts?: DunnoClientOptions) {
    this._dunno = getDefaultClient(opts);
    import('openai').then(({ default: BaseOpenAI }) => {
      this._inner = new BaseOpenAI();
    });
  }

  get chat() {
    const dunno = this._dunno;
    const inner = this._inner;
    return {
      completions: {
        create: async (params: any) => {
          const { dunnoAgent, dunnoSession, dunnoPerson, dunnoAgentVersion, ...rest } = params;
          const start = Date.now();
          const result = await inner.chat.completions.create(rest);
          const latencyMs = Date.now() - start;
          if (dunnoSession) {
            dunno.events.create({
              eventName: 'llm',
              properties: {
                model: result.model,
                inputTokens: result.usage?.prompt_tokens,
                outputTokens: result.usage?.completion_tokens,
                latencyMs,
                messages: rest.messages,
              },
              dunnoSession,
              dunnoAgent,
              dunnoAgentVersion,
              dunnoPerson,
            });
          }
          return result;
        },
      },
    };
  }
}

// Default export: the class
export { OpenAIClass as OpenAI };

File: sdks/python/dunno/ai/provider_openai.py
"""Dunno-wrapped OpenAI client. Requires: pip install dunno[openai]"""
from __future__ import annotations

import time
from typing import Any, Optional

try:
    import openai as _openai
    from openai import OpenAI as _BaseOpenAI
    from openai.types.chat import ChatCompletion
except ImportError as e:
    raise ImportError("Install dunno[openai]: pip install dunno[openai]") from e

from dunno.client import DunnoClient
from dunno.models import EventProperties

_DEFAULT_CLIENT: Optional[DunnoClient] = None


def _get_default_client() -> DunnoClient:
    global _DEFAULT_CLIENT
    if _DEFAULT_CLIENT is None:
        _DEFAULT_CLIENT = DunnoClient()
    return _DEFAULT_CLIENT


class _WrappedCompletions:
    def __init__(self, base: Any, dunno: DunnoClient) -> None:
        self._base = base
        self._dunno = dunno

    def create(
        self,
        *,
        dunno_agent: Optional[str] = None,
        dunno_session: Optional[str] = None,
        dunno_person: Optional[str] = None,
        dunno_agent_version: Optional[str] = None,
        **kwargs: Any,
    ) -> ChatCompletion:
        start = time.monotonic()
        result = self._base.create(**kwargs)
        latency_ms = int((time.monotonic() - start) * 1000)

        if dunno_session:
            usage = getattr(result, "usage", None)
            props = EventProperties(
                model=result.model,
                input_tokens=getattr(usage, "prompt_tokens", None),
                output_tokens=getattr(usage, "completion_tokens", None),
                latency_ms=latency_ms,
                messages=kwargs.get("messages"),
            )
            self._dunno.events.create(
                event_name="llm",
                properties=props,
                session=dunno_session,
                agent=dunno_agent,
                agent_version=dunno_agent_version,
                person=dunno_person,
            )

        return result


class _WrappedChat:
    def __init__(self, base: Any, dunno: DunnoClient) -> None:
        self.completions = _WrappedCompletions(base.completions, dunno)


class OpenAI(_BaseOpenAI):
    """Drop-in OpenAI client that automatically tracks events via Dunno."""

    def __init__(self, *args: Any, dunno_client: Optional[DunnoClient] = None, **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)
        _dunno = dunno_client or _get_default_client()
        self.chat = _WrappedChat(super().chat, _dunno)  # type: ignore[assignment]
